lib/src/layer/polyline_layer.dart

```diff
@@ -39,66 +79,292 @@ class Polyline {
+                          e.onSecondaryTap?.call(lastHit.hit!.point);
+                          return e.tapKey;
+                        })
+                        .whereNotNull()
+                        .toList(),
+                    lastHit.hit!.point,
+                  );
+                },
+                child: paint,
+              ),
+            )
+          : paint,
     );
   }
 }
 
-class PolylinePainter extends CustomPainter {
-  final List<Polyline> polylines;
-
-  final MapCamera map;
+class _PolylinePainter<TapKeyType extends Object> extends CustomPainter {
+  final List<Polyline<TapKeyType>> polylines;
+  final MapCamera camera;
   final LatLngBounds bounds;
+  final _LastHit? lastHit;
+  final double onTapTolerance;
+  final bool tappablesObscure;
+  final bool nonTappablesObscure;
 
-  PolylinePainter(this.polylines, this.map) : bounds = map.visibleBounds;
+  _PolylinePainter({
+    required this.polylines,
+    required this.camera,
+    required this.lastHit,
+    required this.onTapTolerance,
+    required this.tappablesObscure,
+    required this.nonTappablesObscure,
+  }) : bounds = camera.visibleBounds;
 
   int get hash => _hash ??= Object.hashAll(polylines);
 
   int? _hash;
 
+  @override
+  bool? hitTest(Offset position) {
+    if (lastHit == null) return null;
+
+    final hits = <Polyline<TapKeyType>>[];
+    final origin =
+        camera.project(camera.center).toOffset() - camera.size.toOffset() / 2;
+
+    outer:
+    for (final p in polylines.reversed) {
+      // TODO: For efficiency we'd ideally filter by bounding box here. However
+      // we'd need to compute an extended bounding box that accounts account for
+      // the stroke width.
+      // if (!p.boundingBox.contains(touch)) {
+      //   continue;
+      // }
+
+      final offsets = getOffsets(camera, origin, p.points);
+      final strokeWidth = p.useStrokeWidthInMeter
+          ? _metersToStrokeWidth(
+              origin,
+              p.points.first,
+              offsets.first,
+              p.strokeWidth,
+            )
+          : p.strokeWidth;
+      final maxDistance =
+          (strokeWidth / 2 + p.borderStrokeWidth / 2) + onTapTolerance;
+
+      for (int i = 0; i < offsets.length - 1; i++) {
+        final o1 = offsets[i];
+        final o2 = offsets[i + 1];
+
+        final distance = math.sqrt(_distToSegmentSquared(
+          position.dx,
+          position.dy,
+          o1.dx,
+          o1.dy,
+          o2.dx,
+          o2.dy,
+        ));
+
+        if (distance < maxDistance) {
+          if (nonTappablesObscure && p.onTap == null && p.tapKey == null) {
+            break outer;
+          }
+          hits.add(p);
+          if (tappablesObscure) break outer;
+        }
+      }
+    }
+
+    if (hits.isEmpty) return false;
+
+    lastHit!.hit = (
+      // Remove duplicates caused when the hit is close to two segements
+      tapped: LinkedHashSet<Polyline<TapKeyType>>.from(hits).toList(),
```

Comment:
Hit is being called when ever hit detection happens, also on hovers etc. Or  in other words when you move the mouse it's called on every frame. It would be nice to avoid:

1. the deduping as mentioned above
2. and the list allocation. Could you call clear on the pre-existing list instead?

Reply:
The deduping is purely for the reason mentioned above, it does not happen in order to dedupe from non-click hit detections.

I can try using a single list, I'll see if that works.

Reply:
Resolved.